### ISSUE TITLE
Batch normalization

### ISSUE THREAD

**Issue by ikostrikov** [gpapamak/maf#6]
Hi!

Thanks for sharing amazing work!

I'm trying to port your code to PyTorch (for further use in my research).

I have a question regarding your implementation of Batch Norm. As you mention in the paper, it's implemented using global batch statistics. Could you please provide pointers to the lines where it is implemented exactly? My knowledge of Theano is a little bit rusty.

**Comment by gpapamak** [comment]
Thanks!

The `BatchNorm` class has Theano variables `m` and `v` that are the mean and variance calculated on the minibatch, and variables `bm` and `bv` that are stateful Theano variables that contain the mean and variance on the full training set. The model uses `m` and `v` by default.

At the end of training, the full training set is run through the model once and `bm` and `bv` are calculated. This happens here:
https://github.com/gpapamak/maf/blob/3239e801aa6d9ab8d1b804f26a31806cc588018a/ml/trainers.py#L167

Each model has an `eval` function that calculates the log probability at test time. In `eval`, `m` and `v` are substituted with `bm` and `bv`, so that the log probability is calculated with the training batch statistics and not with the test minibatch statistics. As an example, for `MaskedAutoregressiveFlow` this happens here:
https://github.com/gpapamak/maf/blob/3239e801aa6d9ab8d1b804f26a31806cc588018a/ml/models/mafs.py#L97
